### ISSUE TITLE
Control the size and shape of generated shards in voronoi fracture.

### ISSUE THREAD

**Issue by alda30** [scorpion81/blender-fracture#5]
Hello,
I am doing my research on granular systems using Bullet. For this purpose I use voronoi fracture feature of Bullet very frequently. 
I am interested to know if I can control (qualitatively) the shape of shards, or the range of sizes of generated shards. Does anyone know how can do that?

What I do normally is that to use built-in voronoi fracture feature of Bullet, However, I couldn't find any parameter in there to control the size range and shapes of shards. However, when I see the "cell fracture" feature in blender software, I see there are quite nice parameters to control it. I was wondering is the C++ code of the "cell fracture" feature is available? so I can implement it to generate convex hull shapes in bullet in a size range and shapes that I want.

I appreciate your help.
Thanks
Ehsan

**Comment by scorpion81** [comment]
Hi,

the shard shape (note, this will be always voronoi-like unless you intersect it boolean-wise with geometry) and size is controlled by the density of a point cloud basically. The points making up this cloud are particle positions or vertex positions and additionally some more cursor distance based options in cell fracture.
The cell fracture is actually coded in python, you can find the code in your addons directory or take a look here for a quick browse.
This uses bullets convexhull generation, by the way (via a blender operator)

https://developer.blender.org/diffusion/BA/browse/master/object_fracture_cell/;407d0ea752b3af73d3f13ba072671bd09eefecb1

Alternatively, you might consider using the blender fracture modifier build, which implements the fracturing in C with the help of an external library, voro++, and boolean intersection (with the help of blender internal methods) of those shards.
The point cloud sources i chose there are quite similar to cell fracture.

Voro++ can be found here:
http://math.lbl.gov/voro++

It is included (by me) into the blender sources here, with a C interface written by me as well, see here:

https://github.com/scorpion81/blender-fracture/tree/fracture_modifier/extern/voro%2B%2B/src

Point cloud generation happens here, to feed points into the voro++ container later
https://github.com/scorpion81/blender-fracture/blob/fracture_modifier/source/blender/modifiers/intern/MOD_fracture.c#L1036

Fracturing the voro++ container (the boundingbox) happens here:
https://github.com/scorpion81/blender-fracture/blob/fracture_modifier/source/blender/blenkernel/intern/fracture.c#L1481

Basically you add points to a container or domain in voro++ and can query your desired raw data from it, i used this data to build mesh cells:

http://math.lbl.gov/voro++/doc/custom.html

Hope this helps you a bit and points you into the right direction :)
Feel free to ask further questions, if necessary :)

scorpion81

**Comment by alda30** [comment]
Thanks for your comprehensive reply.

In blender when you generate the random point cloud within a volume of an object, without changing the position of points in the point cloud (lets assume random evenly distributed), you can make object shapes biased in any direction by a parameters called "scale cell shape". This is exactly what I want to implement in my code.

Besides, another parameter which changes size of the generated shards is "source noise". By increasing that you can get very large and very small objects. I guess this parameter is applied during random point generation. But how? Any famous distribution is used there?
The normal procedure in Bullet produces same-size shards: https://www.youtube.com/watch?v=_SO-636okrk

---

So conclusion is that I need "scale cell shape" and "source noise". (or any other alternatives which makes the same effect.)

---

I was wondering if the current voronoi shattering code included in Bullet examples (https://github.com/bulletphysics/bullet3/blob/master/examples/VoronoiFracture/VoronoiFractureDemo.cpp) is written by you or not.
But in order to avoid big changes in my code, first lets focus on voronoi shattering feature in bullet (of course in the case you are familiar with that). So if you are familiar that can you help me how can I change them in that code?

In the case you are not familiar with that, can you guide me how can I implement your code in C in my own code. What I need, are:
1) how can I manipulate the parameters above
2) how to determine the shape to be shattered? in current bullet code, only cuboid shapes are available I guess (is that true?)
3) how to get vertices from voronoi shattering to pass them to create a convex hull shape.

Also Another question is that in bullet code, there are two separate functions called "voronoiBBShatter" and "voronoiConvexHullShatter". Do you know what is the difference?

I asked a lot!
Thanks for your kind help.

**Comment by scorpion81** [comment]
Hi, 

regarding the bullet voronoi fracture method, no... i didnt write it, sorry. In fact i used voro++ because it was the easiest way (for me, in this case) to get it done. I struggled myself to get 3D voronoi cells generated from scratch.
And i think the noise ist just a sequence of random numbers applied to some more regular distribution (only guessing, as i also did not write cell fracture)

And regarding voro++...
That

https://github.com/scorpion81/blender-fracture/blob/fracture_modifier/extern/voro%2B%2B/src/c_interface.cc

is the C (to C++) interface i have written, because blenders code is mainly in C and voro++ is in C++.
Starting from here 

https://github.com/scorpion81/blender-fracture/blob/fracture_modifier/source/blender/blenkernel/intern/fracture.c#L1530

the steps are listed... you create a new voronoi container with the size of the bbox of the object (and yes, you can fracture only cuboid objects directly, unless you set "walls", but that is explained on the voro++ homepage and doesnt work for all arbitrary meshes)

http://math.lbl.gov/voro++/examples/

Then you add in the pointcloud points as a sequence of "particles" (just a naming convention of voro++ for the later cell centroids)
Basically now you allocate as many cells as you had points in your cloud (thats only necessary due to the C interface returning values in memory) and invoke the "compute" method.
This creates raw vertex and face data for all cells of the container, usable to build a mesh from.
The parse_cells method in my case builds meshes and later intersects them by boolean cuts against the original geometry.

I think you just need to rip out the "Voro++" dir out of the blender "external" dir, basically you just need to copy the c_interface.cc and c_interface.hh files and add them to the "official" voro++.

That depends on how you want to integrate that lib... either standalone or directly compiled in.

Hope i could help you, although i am not familiar on how bullet does the voronoi fracturing...

Edit: ah regarding "stretched" cells, that is done in blender (by me) by doing tricks with the object transformation (fracturing a scaled copy / container and pointcloud of the object in the background, basically and scaling back, it was a bit complicated and i would need to look it up, dont know off hand now...)

I have no idea whether bullet supports this directly.

scorpion81

**Comment by alda30** [comment]
Thanks for the useful comments. 
I'll be trying to do it, if I have any questions, I'll ask then.

Ehsan

**Comment by alda30** [comment]
Thanks again for the hints.
I succeeded to hack the current vornoi code of Bullet to do the stuff above.
I generated noise by combination of one or multiple normal distributions + a uniform distribution for voronoi points.

In addition, thanks to your hint : 

> Edit: ah regarding "stretched" cells, that is done in blender (by me) by doing tricks with the object >transformation (fracturing a scaled copy / container and pointcloud of the object in the background, >basically and scaling back, it was a bit complicated and i would need to look it up, dont know off >hand now...)

I managed to do the cell stretching.
You can view the output videos:
https://www.youtube.com/watch?v=_SO-636okrk
https://youtu.be/Jf0wUNeCeqU
https://youtu.be/aH212o-InPo

Regards,
Ehsan
